### PR TITLE
Bypass system proxy for local webgui access

### PR DESCRIPTION
**Pull request by ghost**
Purpose
Users (like me) having a web proxy set on their wifi connection or using a pac proxy file cannot access the web UI of syncthing via WebGUIActivity. This PR should fix issue https://github.com/syncthing/syncthing-android/issues/440

Cause
The android webview always uses the globally defined proxy server or PAC file as specified in mobile or wifi connection by default.

Solution
With this implementation, we'll set an exception to bypass the proxy when syncthing app is connecting to localhost (and 127.0.0.1, ...). The web UI then opens properly even if a proxy or PAC file is defined.

Implementation
I noticed that somehow the sent information about bypassing the proxy takes too much time and webview hits the timeout 5 seconds later. To catch this situation, I am receiving the web view's timeout event and then reload the page. This also catches timeouts if the web UI is open and a change in phone's network connectivity occurs. Then, the proxy bypass must be set again.

Compatibility
Backwards compatible, lowest supported version is KitKat.

Testing
Verified working on Android 7.1.2 @ lg-h815.

### CODE REVIEW

File: app/src/main/java/com/nutomic/syncthingandroid/activities/WebGuiActivity.java
Comment (AudriusButkevicius) [inline app/src/main/java/com/nutomic/syncthingandroid/activities/WebGuiActivity.java]: integer.tostring pls

File: app/src/main/java/com/nutomic/syncthingandroid/activities/WebGuiActivity.java
Comment (AudriusButkevicius) [inline app/src/main/java/com/nutomic/syncthingandroid/activities/WebGuiActivity.java]: This can be above meddling with the properties, or potentially just skip the alternative reload

File: app/src/main/java/com/nutomic/syncthingandroid/activities/WebGuiActivity.java
Comment (AudriusButkevicius) [inline app/src/main/java/com/nutomic/syncthingandroid/activities/WebGuiActivity.java]: Is there no non-magical non-relfection way to do this? Where did you find the guide to do this?

Comment (Catfriend1) [reply app/src/main/java/com/nutomic/syncthingandroid/activities/WebGuiActivity.java]: Unfortunately no. After doing deeper research, I am convinced this can only be done using reflection. Source was https://stackoverflow.com/a/26781539

File: app/src/main/java/com/nutomic/syncthingandroid/activities/WebGuiActivity.java
Comment (Catfriend1) [reply app/src/main/java/com/nutomic/syncthingandroid/activities/WebGuiActivity.java]: I verified it working on my phone as I have a proxy in place. With (and without) the proxy set in the android network connection, the exception list gets interpreted correctly. Your link is also valid, but considers using the proxy_config_service - here is my source of information explaining the java side of proxy exceptions that are addressed herein: https://docs.oracle.com/javase/7/docs/api/java/net/doc-files/net-properties.html

File: app/src/main/java/com/nutomic/syncthingandroid/activities/WebGuiActivity.java
Comment (AudriusButkevicius) [inline app/src/main/java/com/nutomic/syncthingandroid/activities/WebGuiActivity.java]: Why the emptu string addition at the end?

File: app/src/main/java/com/nutomic/syncthingandroid/activities/WebGuiActivity.java
Comment (Catfriend1) [reply app/src/main/java/com/nutomic/syncthingandroid/activities/WebGuiActivity.java]: Thats correct. There are also workarounds for older Android versions around, but I think we should not fix the issue with the web UI not loading for that old android versions. It should be rare that a user uses proxy and a very old android version and the fix for ICS would add more complexity we carry with us.

File: app/src/main/java/com/nutomic/syncthingandroid/activities/WebGuiActivity.java
Comment (AudriusButkevicius) [inline app/src/main/java/com/nutomic/syncthingandroid/activities/WebGuiActivity.java]: No point to reload if this returns false

File: app/src/main/java/com/nutomic/syncthingandroid/activities/WebGuiActivity.java
Comment (Catfriend1) [reply app/src/main/java/com/nutomic/syncthingandroid/activities/WebGuiActivity.java]: Put a return after the reload and verified on my phone. It works without calling the base class' error handler after view.reload. Thanks for pointing it out.

**Comment by Catfriend1** [comment]
Note - I verified this working after latest commits.

**Comment by AudriusButkevicius** [comment]
So lgtm, but I'd like @capi or @Nutomic to have a look at this before merging, so lets give them a few days and timeout after that.

File: app/src/main/java/com/nutomic/syncthingandroid/activities/WebGuiActivity.java
Comment (AudriusButkevicius) [reply app/src/main/java/com/nutomic/syncthingandroid/activities/WebGuiActivity.java]: Doesn't this cause an endless loop in case of proxy failure?
If the connection times out, we set the proxy, which succeeds, we reload which then causes a timeout again, etc.

File: app/src/main/java/com/nutomic/syncthingandroid/activities/WebGuiActivity.java
Comment (AudriusButkevicius) [reply app/src/main/java/com/nutomic/syncthingandroid/activities/WebGuiActivity.java]: If we do it here, why do we need it onReceivedError?

**Comment by AudriusButkevicius** [comment]
Apart from the new comments about the endless loop and needing onReceiveError, looks good, let's get this merged, and make a release.

File: app/src/main/java/com/nutomic/syncthingandroid/activities/WebGuiActivity.java
Comment (Catfriend1) [reply app/src/main/java/com/nutomic/syncthingandroid/activities/WebGuiActivity.java]: Thanks for the review. I will test an alternative tomorrow and report back if the endless loop can be improved.

Comment (AudriusButkevicius) [reply app/src/main/java/com/nutomic/syncthingandroid/activities/WebGuiActivity.java]: Isn't it enough just to set it on the web view upon construction?

Comment (Catfriend1) [reply app/src/main/java/com/nutomic/syncthingandroid/activities/WebGuiActivity.java]: Unfortunately no, tested this before and found it not working by a chance of ~80%. Thats why I put the reload and setproxy on timeout a second time. Maybe there is some delay in the app to webview "communication" or delay on the webview handling the updated proxy settings. Im sure we find a solution on that...

Comment (AudriusButkevicius) [reply app/src/main/java/com/nutomic/syncthingandroid/activities/WebGuiActivity.java]: 80% chance on what. This is not quantum computing, our code is deterministic.

**Comment by Catfriend1** [comment]
@AudriusButkevicius Ok, I fixed it. There were two problems in the code. You first need to set domStorageEnabled (https://stackoverflow.com/questions/18377769/webview-not-able-to-load-https-url-in-android) and then set the proxy properly. If one is missing, the WebGUI does not load when a PAC file/proxy is configured for the network. lgtm?

**Review by capi — APPROVED** [review]
I have only reviewed the diff and besides my dislike of the reflection magic that has been noted by @AudriusButkevicius already, I don't see any huge issues.

But this review is based only by looking at the diff, I didn't try it yet. And I'm unfortunately not too familiar with WebView and how to properly configure it.

**Comment by Nutomic** [comment]
The main problem with beta releases is that F-Droid doesn't support beta releases automatically. So you have to bug F-Droid maintainers to mark the release as beta, or every F-Droid user will get the beta release as automatic update. At least that's how it was last year, but maybe this was improved in the mean time.

Alternatively, you could publish the beta on Google Play, but not push the tag to Github so F-Droid doesn't pick it up.

**Pull request merged**
